# Programming Python

## Philosophy Python

The developers of the Python language adhere to a certain programming philosophy,
called "The Zen of Python" ("Zen Python", or "Zen Paiton"). Her text is rendered Python interpreter by the import this command (works once per session). 

In general, she is suitable for programming in any language.
- Text philosophy
- Beautiful is better than ugly.
- Explicit is better than implicit.
- Complicated is better than confusing.
- Flat is better than nested.
- Sparse is better than tight.
- Readability matters.
- Special cases are not so special that they break the rules.
- Practicality is more important than flawlessness.
- Mistakes should never be glossed over.
- If they are not glossed over explicit.
- When faced with ambiguity, reject the temptation to guess.
- There must be one and, preferably, only one obvious way to do this.
- Although it may not be obvious at first, if you aren't Dutch.
- Now is better than never.
- Though never often better than right now.
- If implemenatation is difficult to explain - it's a bad idea.
- If the implementation is easy to explain - it's idea may be good.
- Namespace is great thing.


In [1]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


## Python as calculater

In [3]:
4*7 # multiplication, answer is 28

28

In [5]:
3*(2+5) # operator priority

21

In [6]:
5**3 # exponentiation

125

With adding, multiplication and exponentiation be understandable. At first sight division also not foreshadows problems.


In [7]:
12/3 # operator division

4.0

**Attention!** The behaviour of Python version 2 was different: he always divided by an integer (discarding the remainder).This led to a lot of problems when the developer expected the division to be normal rather than integer.In a Python version 3 division is in floating number;if you want a integer you can use double slash( // ).

In [8]:
3//2 # integer division

1

How are the other operators?Let's try to extract the square root.

In [10]:
sqtr(4) # not work

NameError: name 'sqtr' is not defined

This function can be found in the standart module *math*. To use it, you need to import this module.It can be done in different ways.

In [11]:
import math
math.sqrt(4)

2.0

Once the math module has been imported, you can find out what other functions it contains.В IPython Notebook to do this, simply enter the module name, put a dot and press the "Tab".

For example, sinus :

In [13]:
math.sin(0)

0.0

The given syntax may be awkward if you often have to call any mathematical functions. To avoid having to write the word "math" every time, you can import from the module specific functions.

In [16]:
from math import sqrt
sqrt(4)

2.0

## Float numbers and uncertainty

Float numbers are not so simply. Here, for example, let's calculate the sine of the number π:


In [17]:
from math import pi,sin
sin(pi)

1.2246467991473532e-16

Unclear answer?At the first,it's the so-called a computer form exponential number note. She comfortable, if need can write so very big or so small numbers. 1.2E2 means 1,2 * 10^2, that is 1200, and 2.4e-3 that is that 2,4 * 10^(-3) = 0,0021. Result it is a calculated by Python for sine pi have been order 10^-16 , that is so small number that close null. Why is't "true" null?In to float number all calculated doing to be computer with to some limited accuracy what often true answer make approximate.To be prepared for that.

## Variables

In [20]:
x = 4 #dynamic typing
x # output currecny variable x

4

In a Python should't declare variable - only enough assign she's value and it may used.

In [21]:
x + 2 # assign

6

In [22]:
x = x +2 # reassign

In [23]:
x # console output

6

Variables can store not just numbers,but and other data.Example, strings.We will come across this in the next section.

## Input data

Work to Jupyter/ IPython Notebook rarely requires write code what self to self request data from keyboard, but for other application (and particular for homeworks) is may be. Also, writing interactive applications in the form of such a fun session. Writing ,for example, application which says hello us by name.

In [25]:
name = input("Input your name: ")
print("Hello, ",name)

Input your name: Daniil
Hello,  Daniil


In [27]:
name

'Daniil'

Try now write app "doubler".She's should input number, doubler he and return result.

In [28]:
x = input("Input any number : ") #return str
print(x*2)

Input any number : 3
33


What is wrong?Of course, in a way, the computer did what we asked for - there was one number and now there are two, but we wanted something different.Why is this so?The point is what input always return string(he don't known what we will input - number or anyway characters; he don't what in parameter said " Input anyway number" - for her is a only some string ).And in variable right now is no number 12, but string '12' (Take note on apostrophes).

In [29]:
x

'3'

So what to work her with number need turn from string type to numberic.Example,if we want work only with integer numbers then suitable will be int(from word integer , full number).

In [30]:
int(x)

3

See, no apostrophes - it's a number now. Word int it's simultaneously and term type date (integer numbers) and func , which she's turn in what handed over in parameter an indicated type.

Rewritting app doubler as follows

In [33]:
x_str = input("Input any number: ")
x = int(x_str)
print(x*2)

Input any number: 3
6


It's a different story! Now in x_str lies string which us inputed, in x place result she's convert in integer number and next this number double and output on screen. Would not have to be output more variable(saving a line).


In [34]:
x = int(input("Input any number: "))
print(x*2)

Input any number: 41
82


To try again.

In [35]:
x = int(input("Input any number: "))
print(x*2)

Input any number: 12.34


ValueError: invalid literal for int() with base 10: '12.34'

I tryed input number 12.34 (no integer) and Python did't can cite her to integer type ( that's can from string "12.34" integer num and issued error).If we want work with no integer number we was need should type float.

In [36]:
x = float(input("Input any number (not necessarily integer): "))
print(x*2)

Input any number (not necessarily integer): 12.3
24.6


It all works!

By the way, you can use the str function to turn a number into a string. Generally, the name of any data type, when called as a function, attempts to convert it's argument into an appropriate data type.

## Practical exercise: Fibonacci numbers

*Fibonacci numbers* are a sequence of numbers starting with two ones and such that the next number in the sequence is equal to the sum of the previous two.

a1 = 1;
a2 = 1;
an+1 = an + an-1 for all n > 2

Example, a3 = 1 + 1 = 2, a4 = 2 + 1 = 3

**Task** : count 15th fibonacci number

In [39]:
a = 1 # first number
b = 1 # second number
i = 2 # position of that number which are place in variable b (now there a_2)

We will be store two last num, because exactly they define next, as well as position last finded num (var i).

In [45]:
c = a + b # finded next num
i = i + 1 # enlarge i on 1
a = b # value a not actually, but value b will still come in handy
b = c # remember calculated value
print(i,b)

8 21


Python performs commands consistently, string to string, so position commands important.Performing this cell repeatedly,you will be get another next number Fibonacci.

**Test question.** What happens if we swap the two lines until print?

In [47]:
# Repeat this code again
c = a + b # finded next num
i = i + 1 # enlarge i on 1
a = b # value a not actually, but value b will still come in handy
b = c # remember calculated value
print(i,b)

9 34


In [48]:
# And again
c = a + b # finded next num
i = i + 1 # enlarge i on 1
a = b # value a not actually, but value b will still come in handy
b = c # remember calculated value
print(i,b)


10 55


In [49]:
# And again
c = a + b # finded next num
i = i + 1 # enlarge i on 1
a = b # value a not actually, but value b will still come in handy
b = c # remember calculated value
print(i,b)


11 89


Tip.Didn't have to copy this cell with code many round, but simple starting the same cell many round.Tip.Didn't have to copy this cell with code many round, but simple starting the same cell many round.The easiest thing to do with combination Ctrl+Enter. And, of course, on practic we will not manually perform same code many round - for this is cicles. But we get to know on next lection.